libraries/AP_OpenDroneID/AP_OpenDroneID.cpp

```diff
@@ -95,7 +103,81 @@ void AP_OpenDroneID::init()
         return;
     }
 
+    load_UAS_ID_from_persistent_memory();
     _chan = mavlink_channel_t(gcs().get_channel_from_port_number(_mav_port));
+    _initialised = true;
+}
+
+void AP_OpenDroneID::load_UAS_ID_from_persistent_memory()
+{
+    id_len = sizeof(id_str);
+    size_t id_type_len = sizeof(id_type);
+    size_t ua_type_len = sizeof(ua_type);
+    if (hal.util->get_persistent_param_by_name("DID_UAS_ID", id_str, id_len) &&
+        hal.util->get_persistent_param_by_name("DID_UAS_ID_TYPE", id_type, id_type_len) &&
+        hal.util->get_persistent_param_by_name("DID_UA_TYPE", ua_type, ua_type_len)) {
+        if (id_len && id_type_len && ua_type_len) {
+            _options.set_and_save(_options.get() & ~LockUASIDOnFirstBasicIDRx);
+            _options.notify();
+            GCS_SEND_TEXT(MAV_SEVERITY_INFO, "OpenDroneID: Locked UAS_ID: %s", id_str);
+        }
+    } else {
+        id_len = 0;
+    }
+}
+
+void AP_OpenDroneID::set_basic_id() {
+    if (pkt_basic_id.id_type != MAV_ODID_ID_TYPE_NONE) {
+        return;
+    }
+    if (id_len == 0) {
+        load_UAS_ID_from_persistent_memory();
+    }
+#if defined(OPENDRONEID_UA_MFR_CODE)
+    if ((_options & UseChipIDAsBasicID) && (id_len == 0)) {
+        float val;
+        // prepare basic id pkt
+        AP_Param::get("SYSID_THISMAV", val);
+        pkt_basic_id.target_system = val;
+        pkt_basic_id.target_component = MAV_COMP_ID_ODID_TXRX_1;
+        pkt_basic_id.id_type = MAV_ODID_ID_TYPE_SERIAL_NUMBER;
+        pkt_basic_id.ua_type = OPENDRONEID_UA_TYPE;
+
+        uint8_t sys_id[12];
+        uint8_t len = 12;
+        hal.util->get_system_id_unformatted(sys_id, len);
```

Comment:
this assumes stm32 ID layout, not true on linux hal or esp32, or SITL
maybe use crc_crc64() ?

Reply:
or remove feature is fine too

Reply:
removed